feat: Add workflow dependencies visualization (#26062)

xDarkmanx opened the pull request:
## feat: Add workflow dependencies visualization

Related to #26062

### What does this PR do?

This PR adds an interactive visualization component that displays job dependencies in Gitea Actions workflow runs. It helps users understand complex pipeline structures at a glance, addressing the difficulty of comprehending dependency chains in current Gitea UI.

### Screenshots / Images

#### Example 1: Multi-level CI/CD Pipeline (9 levels deep)

![Multi-Level Pipeline](https://gitea.dev.koenig.su/wf_test1/wf_test1.png)
[Pipeline Source](https://gitea.dev.koenig.su/wf_test1/wf_test1.yml)

#### Example 2: Microservices Deployment with Matrix Jobs

![Microservices Pipeline](https://gitea.dev.koenig.su/wf_test2/wf_test2.png)
[Pipeline Source](https://gitea.dev.koenig.su/wf_test2/wf_test2.yml)

#### Example 3: Real production workflow

![Production Example](https://gitea.dev.koenig.su/wf_prod/wf_prog.png)
[Pipeline Source](https://gitea.dev.koenig.su/wf_prod/wf_prog.yml)

### Docker Image for Testing

A pre-built Docker image with these changes is available for reviewers to test without building locally:
```bash
docker pull xvulturex/gitea:latest
docker run -p 3000:3000 xvulturex/gitea:latest
```

### What changes were made?

#### API Changes:

- Added `job_id` field to `structs.ActionJob` for YAML job identifier mapping
- Added `needs` field to track job dependencies as defined in workflow YAML
- Both fields are `omitempty` for backward compatibility
- Updated Swagger documentation for new fields

#### Frontend Component:

- New Vue 3 component `WorkflowGraph.vue` with TypeScript support
- Interactive features: zoom with mouse wheel, pan by dragging, hover highlighting
- Visual status indicators with animations (success, failure, running, etc.)
- Support for matrix jobs through `job_id` field correlation
- Responsive design using CSS variables for theme compatibility
- Click navigation to job details page

#### Backend:

- No database schema changes required
- Uses existing columns (`job_id` maps to existing `job_name` in actions table)
- Handles edge cases: cycles, missing dependencies, invalid job references

### Testing

The implementation has been thoroughly tested with:

#### Manual Testing:

- ✅ Simple workflows (3-5 jobs)
- ✅ Complex dependency chains (20-50+ jobs, 9+ levels deep)
- ✅ Matrix job expansions (6x2, 3x3 matrices)
- ✅ Edge cases (cycles, missing dependencies handled gracefully)
- ✅ Theme compatibility (light and dark modes)
- ✅ Mobile responsive behavior

#### Test Workflows Used:

1. **Multi-Level CI/CD Pipeline** - 9 dependency levels, parallel branches, matrix jobs
2. **Microservices Deployment** - Complex matrix jobs, multiple dependency paths
3. **Real production workflow** - Validated behavior with actual user workflow

#### Performance:

- Handles workflows with 20-50 jobs efficiently
- Uses requestAnimationFrame for smooth zoom/pan animations
- Optimized SVG rendering with minimal DOM updates

### Backwards Compatibility

- ✅ **Full backwards compatibility** - New API fields are optional (`omitempty`)
- ✅ **No breaking changes** - Existing API responses unchanged
- ✅ **Optional feature** - Visualization only shown when dependency data is available
- ✅ **Fallback behavior** - No graph shown if no dependencies exist
- ✅ **Database compatibility** - No schema migrations required

### Usage

1. **Visualization appears automatically** when viewing workflow runs with dependencies
2. **Interactive controls**:
   - Zoom: Mouse wheel or +/- buttons
   - Pan: Click and drag
   - Reset: Reset view button
3. **Navigation**: Click any job node to navigate to its details page
4. **Visual indicators**:
   - Green: Success/Completed
   - Red: Failed
   - Yellow: Running
   - Gray: Waiting/Blocked

### Checklist

- [x] Code follows the Gitea code style guidelines
- [x] Self-reviewed the code changes
- [x] Added necessary documentation (Swagger updates)
- [x] Backwards compatible changes
- [x] Considered performance implications
- [x] Tested on mobile/responsive views
- [x] Used CSS variables for theming
- [x] No console errors in development mode

### Additional Notes

- The graph uses a topological layout algorithm to arrange jobs by dependency levels
- Matrix jobs are properly handled via the `job_id` field which maps to YAML job identifiers
- All styling uses CSS variables (`var(--color-*)`) for proper theme integration
- The component is designed to be maintainable with clear separation of concerns
- Handles edge cases with appropriate warnings in console

xDarkmanx committed:
feat: Add workflow dependencies visualization

    ## API changes:
    - Add job_id field for YAML job identifiers
    - Add needs field for job dependencies
    - Update swagger documentation

    ## Frontend features:
    - Interactive graph with zoom/pan navigation
    - Matrix jobs visualization
    - Custom job names support via job_id
    - Status-based coloring and animations
    - Hover highlighting for related jobs
    - Click to navigate to job details

    ## Technical details:
    - Uses existing database columns (no schema changes)
    - Backward compatible (omitempty for new fields)
    - Handles edge cases (cycles, missing dependencies)
    - Performance optimized

    ## UI integration:
    - Updated RepoActionView.vue styles for proper theme compatibility
    - Consistent styling with Gitea's CSS variables

xDarkmanx committed:
fix: Resolve all linting errors with correct CSS variables

ChristopherHX commented on services/convert/convert.go:
```diff
@@ -367,7 +367,8 @@ func ToActionWorkflowJob(ctx context.Context, repo *repo_model.Repository, task
 
 	return &api.ActionWorkflowJob{
-		ID: job.ID,
+		ID:    job.ID,
```
Why? change rest api

xDarkmanx commented on templates/swagger/v1_json.tmpl:
Yes, they are affected.
However, these changes are backward compatible.
They are only required for the Dependency Graph functionality.

ChristopherHX commented on modules/structs/repo_actions.go:
```diff
@@ -161,6 +161,7 @@ type ActionWorkflowStep struct {
 // ActionWorkflowJob represents a WorkflowJob
 type ActionWorkflowJob struct {
 	ID         int64                 `json:"id"`
+	JobID      string                `json:"job_id,omitempty"`
```
> The needs and job_id fields are required for the workflow dependency graph functionality implemented 

Yes I agree that you need those information to build the graph, but you do not use the gitea rest api structs in your vue component as far as I understood.

Your template uses the data from the view model

Changes of this file
https://github.com/go-gitea/gitea/blob/f685d5f8d586f3970820e5471706fe9b08d3b315/routers/web/repo/actions/view.go

This then directly use ActionRunJob to create this data for your vue component.

xDarkmanx committed:
Merge branch 'main' into feature/workflow-graph

xDarkmanx committed:
Merge branch 'main' into feature/workflow-graph

lautriva commented:
Would it be possible to show the graph "left-to-right" (like Github / Gitlab) instead of "top to bottom"?

xDarkmanx commented:
> Would it be possible to show the graph "left-to-right" (like Github / Gitlab) instead of "top to bottom"?

Good suggestion! I can add an orientation toggle (top-down ↔ left-right) in a separate PR to keep this one focused on the core functionality.

xDarkmanx committed:
Merge branch 'main' into feature/workflow-graph

bircni commented:
@xDarkmanx I would really love to see this - i think its pretty useful

xDarkmanx commented:
> I would really love to see this - i think its pretty useful

I completely agree with you.

You can already try this functionality.

```sh
docker pull xvulturex/gitea:latest
docker run -p 3000:3000 xvulturex/gitea:latest
```

xDarkmanx committed:
Merge branch 'main' into feature/workflow-graph

Signed-off-by: Semenets V. Pavel <p.semenets@gmail.com>

lunny commented:
Looks amazing, I need some time to review it.

xDarkmanx committed:
Merge branch 'main' into feature/workflow-graph

xDarkmanx committed:
Merge branch 'main' into feature/workflow-graph

xDarkmanx committed:
Merge branch 'main' into feature/workflow-graph

xDarkmanx committed:
Merge branch 'main' into feature/workflow-graph

xDarkmanx committed:
Merge branch 'main' into feature/workflow-graph

xDarkmanx committed:
Merge branch 'main' into feature/workflow-graph

xDarkmanx commented:
> The button label “Summary” is a bit confusing—could we consider a different name? Also, should the switch setting be remembered per run or per user? Personally, I think the graph should be hidden by default rather than displayed automatically. Clicking a rectangle in the graph refreshes the page, which results in a poor user experience.

@lunny Thank you for your feedback.

1.  Change the button label (I suggest "Dependency Graph").
2. Save the state per session.
3. The graph will be hidden by default.
4. Clicking a graph node will not refresh the page.

If I understood you correctly, I'm ready to implement these changes.

xDarkmanx committed:
Merge branch 'main' into feature/workflow-graph

bircni commented:
> > The button label “Summary” is a bit confusing—could we consider a different name? Also, should the switch setting be remembered per run or per user? Personally, I think the graph should be hidden by default rather than displayed automatically. Clicking a rectangle in the graph refreshes the page, which results in a poor user experience.
> 
> @lunny Thank you for your feedback.
> 
> 1. Change the button label (I suggest "Dependency Graph").
> 2. Save the state per session.
> 3. The graph will be hidden by default.
> 4. Clicking a graph node will not refresh the page.
> 
> If I understood you correctly, I'm ready to implement these changes.

1. I think summary fits better but I am also happy with Dependency Graph but would prefer "Graph"
2. definitely
3. How is Github doing this?
4. perfect

xDarkmanx committed:
Merge branch 'main' into feature/workflow-graph

xDarkmanx committed:
Merge branch 'main' into feature/workflow-graph

xDarkmanx committed:
Merge branch 'main' into feature/workflow-graph

xDarkmanx committed:
Merge branch 'main' into feature/workflow-graph

xDarkmanx committed:
Merge branch 'main' into feature/workflow-graph

xDarkmanx committed:
fix: add dependency graph toggle with persistent state

- Changed button label to "Dependency Graph" for clarity
- Implemented session-persistent state using localStorage for graph visibility
- Set graph to be hidden by default on page load
- Note: Cannot prevent page reload on graph node click as RepoActionView component
  performs full page navigation when switching between jobs

xDarkmanx commented:
@lunny @bircni 
Colleagues, what's next?
Are there any other improvements that need to be made?

xDarkmanx committed:
Merge branch 'main' into feature/workflow-graph

Signed-off-by: Semenets V. Pavel <p.semenets@gmail.com>

xDarkmanx committed:
Merge branch 'main' into feature/workflow-graph

Signed-off-by: Semenets V. Pavel <p.semenets@gmail.com>

xDarkmanx committed:
Merge branch 'main' into feature/workflow-graph

xDarkmanx committed:
fix: move dependency graph button to header actions group and improve UX

- Moved "Dependency Graph" button to the right side of the header next to "Rerun all jobs" button
- Grouped all action buttons (approve, cancel, rerun, graph) in a consistent UI pattern
- Added locale support for "Dependency Graph" button text across all languages
- Fixed button behavior by using 'primary' class instead of 'link-action' to prevent page reloads
- Added conditional display (only shown when run.jobs.length > 1) to avoid confusion when there's only one job
- Improved user experience by ensuring the graph button is only visible when it's actually useful

This change addresses the confusing behavior where clicking the graph button did nothing when there was only one job, and makes the UI more intuitive by placing related actions together in the header.

xDarkmanx committed:
Merge branch 'main' into feature/workflow-graph

xDarkmanx committed:
Merge branch 'main' into feature/workflow-graph

xDarkmanx committed:
fix: Revert language changes

xDarkmanx commented:
@lunny 
What's next ?

wxiaoguang commented on web_src/js/components/RepoActionView.vue:
```diff
@@ -585,7 +600,16 @@ export default defineComponent({
+        <WorkflowGraph
+          v-if="showSummary && run.jobs.length > 1"
+          :jobs="run.jobs"
+          :current-job-id="parseInt(jobIndex)"
```
`job index` is not `job id`, IIRC

xDarkmanx replied:
Fixed

wxiaoguang commented:
UI problems:

<details>

<summary>overlapped text</summary>
<img width="1062" height="272" alt="image" src="https://github.com/user-attachments/assets/50443007-0d1b-4ef4-959d-cde6cfa78f4e" />

<img width="1060" height="354" alt="image" src="https://github.com/user-attachments/assets/411180bf-85ed-4d86-b375-d727e2964cae" />

</details>


<details>

<summary> strange borders </summary>

<img width="2302" height="1252" alt="image" src="https://github.com/user-attachments/assets/de84fa64-6754-4ebc-926e-f2c295468183" />

</details>

wxiaoguang commented:
> > It feels laggy when dragging. Maybe you can extract the "svg zoom & drag" code from "Add viewer controller for mermaid (zoom, drag) #36557" & "Fix markup code block layout #36578" and reuse them.
> 
> I think I've fixed the lag when dragging the graph.
> 
> At least on my computer, I don't see any lag when zooming or dragging.

Hmm, the laggy is caused by `transition: transform 0.1s ease`. There is a 0.1 second delay between the mouse move and the svg shape's move. That's why I can feel it.

wxiaoguang commented:
> > Another problem of "dragging" is that if the mouse moves out of the svg's container, then the dragging stops, unable to continue dragging the svg back.
> 
> Can I have you write me a task list of what I need to fix? And I will fix it.

At the moment I only did a quick review, some brief thoughts:

1. Use explicit and clear var names for different concepts (e.g.: job id vs job index)
2. Some UI problems: https://github.com/go-gitea/gitea/pull/36248#issuecomment-3931311670
3. Dragging problems: 
    * Feels laggy
    * Loses tracking when mouse moves out of the container
    * Reference: https://github.com/go-gitea/gitea/pull/36248#issuecomment-3931330762

By the way, there is a mock page http://localhost:3000/devtest/repo-action-view , then you can debug the UI without really running an Actions task.

xDarkmanx commented:
> > > Another problem of "dragging" is that if the mouse moves out of the svg's container, then the dragging stops, unable to continue dragging the svg back.
> > 
> > 
> > Can I have you write me a task list of what I need to fix? And I will fix it.
> 
> At the moment I only did a quick review, some brief thoughts:
> 
> 1. Use explicit and clear var names for different concepts (e.g.: job id vs job index)
> 2. Some UI problems: [feat: Add workflow dependencies visualization (#26062) #36248 (comment)](https://github.com/go-gitea/gitea/pull/36248#issuecomment-3931311670)
> 3. Dragging problems:
>    
>    * Feels laggy
>    * Loses tracking when mouse moves out of the container
>    * Reference: [feat: Add workflow dependencies visualization (#26062) #36248 (comment)](https://github.com/go-gitea/gitea/pull/36248#issuecomment-3931330762)
> 
> By the way, there is a mock page http://localhost:3000/devtest/repo-action-view , then you can debug the UI without really running an Actions task.

Another question

Do you want me to use your controller from PR #36557 and #36578 ?

xDarkmanx commented:
> > Another question
> > Do you want me to use your controller from PR #36557 and #36578 ?
> 
> It's up to you. I think #36557 and #36578 already have a good and simple "controller" and many edge cases have been handled.

I need some time to study this code.

xDarkmanx committed:
fix: Improve RepoActionView console styling and WorkflowGraph UX

RepoActionView:
- Restore original console styles in RepoActionView.vue
- Fix light theme console colors (variables were hardcoded to dark values)
- Improve overall UI appearance and readability

WorkflowGraph:
- Remove artificial delay during graph panning
- Fix graph movement behavior
- Add persistence for zoom/position state when switching between jobs
- Rename prop `current-job-id` to `current-job-idx` for accuracy
- Remove border from graph container

xDarkmanx committed:
fix: address code review feedback for WorkflowGraph.vue

Changes:
- i18n: Removed all comments, made code self-documenting
- fix(zoom): Corrected cumulative scaling drift (1.2 * 0.8 != 1.0)
- refactor(styles): Replaced inline dynamic styles with CSS classes
- fix(storage): Implemented LRU cache for localStorage (max 15 entries)
- cleanup: Removed CSS variable fallbacks, use design system only
- fix: Corrected zoom controls rendering and border display
- chore: Aligned component structure with project conventions

Note: Component implemented manually with strict typing and performance
optimizations (rAF, cycle detection)

wxiaoguang commented:
Thank you very much! I think I can help to make some changes later -- yep, every project has its own "frameworks" or "libraries", I can help to use the existing functions like `SvgIcon`, etc.

By the way, if you don't mind, I think I will simplify the UI a little, remove some unnecessary elements like "zoom ratio" , "parallelism" and "legend" from the UI. The reasons that I think we'd better keep the UI simple are:

* end user doesn't really need to know the current "zoom ratio", no matter 150% or 200%, as long as they can see what they need, then it is good enough.
* it seems that "parallelism" doesn't provide enough useful information, the max nodes count can be counted easily at first glance, and IMO parallelism=8 or parallelism=9 doesn't differ
*  "legend" seems redundant, there are already enough information including texts on the UI.


What do you think?

bircni commented:
@wxiaoguang I am fully with you
zoom ratio is not really needed, the other parts also 

<img width="696" height="161" alt="grafik" src="https://github.com/user-attachments/assets/81fb913b-be83-4b1d-a95f-d1e08e18d656" />

- Are these really the same colors as in the other status colors?
- I think the progress bar on the "in progress" is too much

I think a more minimalistic UI as Github is using would look better (no full colored boards)
<img width="627" height="142" alt="grafik" src="https://github.com/user-attachments/assets/f0e3befb-750e-4006-aa5b-e0e9578787c6" />

Also just thinking about - could we get rid of the dependency graph button and instead implement this "Summary" View
would be great as there is another feature request where we want to show summaries of some runs
<img width="323" height="345" alt="grafik" src="https://github.com/user-attachments/assets/8a916620-622f-4530-97b3-4b64a5aafab8" />

xDarkmanx committed:
Merge branch 'main' into feature/workflow-graph

lunny commented on web_src/js/components/RepoActionView.vue:
```diff
@@ -107,6 +110,7 @@ function isLogElementInViewport(el: Element, {extraViewPortHeight}={extraViewPor
 type LocaleStorageOptions = {
   autoScroll: boolean;
   expandRunning: boolean;
+  showSummary: boolean;
```
showWorkflowGraph?

xDarkmanx replied:
Fixed

xDarkmanx committed:
Merge branch 'main' into feature/workflow-graph

xDarkmanx committed:
fix: apply review feedback for workflow graph

- Rename showSummary -> showWorkflowGraph
- Rename constants to camelCase: STORAGE_KEY -> storageKey, MAX_STORED_STATES -> maxStoredStates
- Update button label: "Dependency Graph" -> "Workflow Graph"
- Update locale key: dependencyGraph -> workflowGraph
- Fix node colors by using theme variables (-dark-2) instead of custom fallbacks
- Remove unused CSS styles

xDarkmanx commented on web_src/js/components/RepoActionView.vue:
```diff
@@ -691,6 +719,34 @@ export default defineComponent({
   overflow-wrap: anywhere;
 }
 
+.summary-toggle {
```
Removed

xDarkmanx commented on web_src/js/features/repo-actions.ts:
```diff
@@ -22,6 +22,7 @@ export function initRepositoryActionView() {
       scheduled: el.getAttribute('data-locale-runs-scheduled'),
       commit: el.getAttribute('data-locale-runs-commit'),
       pushedBy: el.getAttribute('data-locale-runs-pushed-by'),
+      dependencyGraph: el.getAttribute('data-locale-runs-dependency-graph'),
```
Fixed

xDarkmanx commented:
Need some time for build new container
ryzen 5 gen 2 !== ryzen 7 gen 5 :(

My home pc is much slower than my work pc.

xDarkmanx committed:
fix: change variable name storegeKey -> storageKey

wxiaoguang committed:
refactor

wxiaoguang commented:
Made some changes.

1. These UI elements are removed, because without them, the UI is still clear enough.

Redundant UI elements only make users spend more time to read.

<details>

<img width="2138" height="1208" alt="image" src="https://github.com/user-attachments/assets/b0d9665f-c106-407c-a5c3-b3066d69886e" />

</details>

2. Use SvgIcon for buttons and fine-tune some CSS styles
3. "mousewheel to zoom" is removed. It only interrupts user's scrolling. Mouse wheel can't be used for zooming in many cases.


----


TODO:

It seems that some edges are not displayed (not caused by my code, old code also doesn't display it):

<img width="1000" height="938" alt="image" src="https://github.com/user-attachments/assets/86ac4407-c894-40d3-9442-39a5b91742b3" />

wxiaoguang committed:
refactor

wxiaoguang committed:
refactor

wxiaoguang committed:
revert api change

wxiaoguang committed:
fix types

wxiaoguang committed:
revert unnecessary change

wxiaoguang committed:
fix prop, fix status

wxiaoguang committed:
fix

wxiaoguang committed:
fix

wxiaoguang committed:
fix dragging & click

wxiaoguang committed:
fix ui

wxiaoguang committed:
fix "needs"

wxiaoguang committed:
Merge branch 'main' into feature/workflow-graph

wxiaoguang committed:
fine tune

wxiaoguang commented on web_src/js/components/WorkflowGraph.vue:
```diff
@@ -0,0 +1,1326 @@
+<template>
+  <div class="workflow-graph" v-if="jobs.length > 0">
+    <div class="graph-header">
+      <h4 class="graph-title">Workflow Dependencies</h4>
```
Maybe leave it to future
